413 - Update token validation flow

nozomione opened the pull request:
## Issue Number

Closes #413

## Purpose/Implementation Notes

I've implemented the user token validation flow during the initial load of the application, which replaces the `useToken` hook, and adjusted the codebase accordingly. 
 
Changes include:
- Added a new `acceptedTerms`context state to track whether the user has previously accepted the terms, which is used to determine if validation is necessary during the application load
- Removed the token validation check from the codebase
- Adjusted the visibility of the terms of use checkbox in the download forms

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

nozomione commented:
I've updated the PR based on the 1:1 feedback, and it's ready for another review. Thank you, David!

Changes include:
- In `RefinebioContext`, added:
  - The application-level token on initial load
  - A side effect for syncing the token state with `acceptTerms`
- Adjusted the consumer hooks (removed token generation i.e., `createToken`) and components (updated form submission and visibility of the terms of use checkbox)
- Renamed the helper and interface method parameter from `params` to `body`

davidsmejia reviewed (commented):
Two changes for now:
- only use token as a dependency of useEffect
- acceptedTerms should be used for component logic
- use `setAcceptedTerms` and remove `applyAcceptedTerms`

nozomione commented:
I've applied the following changes:

>     * only use token as a dependency of useEffect
>     * acceptedTerms should be used for component logic
>     * use `setAcceptedTerms` and remove `applyAcceptedTerms`

To ensure the `token` state is used only within hooks and not in components, I've created a temporary `useTriggerSubmit` hook that triggers the `submit` method in each consumer component when both `token` and the required condition (e.g., download options) are ready.

This PR is ready for another look. Thank you, David!

nozomione commented:
David, thank you for this great idea for the `useWaitFor` hook! 🚀

We discussed using the `useWaitFor` hook to ensure that a callback is only called once a specified condition is met. In our case, certain async methods need to be executed only after a valid token is ready (which is fetched asynchronously via `RefinebioContext`).

```js
import { useEffect, useRef } from 'react'

export const useWaitFor = (condition) => {
  const resolveRef = useRef()
  const promiseRef = useRef(new Promise((resolve) => {
    resolveRef.current = resolve
  }));

  useEffect(() => {
    if (condition) resolveRef.current() // resolve only when condition is met
  }, [condition])
  
  // once the promise is successfully resolved, call the callback
  return (cb) => async (...args) => promiseRef.current.then(cb(...args))
}
```

**Usage:**
1. Instantiate the hook at the `RefinebioContext` provider level and define the `waitForToken` method:
```js
const waitForToken = useWaitFor(validToken) 
```
2. Wrap any methods that require a valid token with `waitForToken` 
```js
const downloadDataset = waitForToken(async (...) => {...}) 
```

Based on this insight, I've updated the codebase. I also customized the hook to initialize only once (for re-renders). Additionally, the hook now passes the latest condition to the callback as the last parameter after resolution, and I've adjusted the method signatures accordingly.

This PR is ready for another look. Thank you David!

nozomione commented:
I've applied the 1:1 feedback, replaced `useWaitFor` with `usePromise`, and made the following adjustments to the codebase:
- Handled form submission using side effects within the components
- Threw an error if `acceptedTerms` is `false` in the following `useDatasetManager` methods:
   - `downloadDataset`
   - `startProcessingDataset`
- Removed the method definitions for creating and activating tokens, and handled them directly within `useEffect` blocks in `RefinebioContextProvider`

This PR is ready for your review. Thank you, David!

davidsmejia reviewed (commented):
This is almost there, just left a few comments.

davidsmejia reviewed (approved):
LGTM
